File: lin-compiler/lin_compiler/Parameter.cs
using System;
using System.Collections.Generic;

namespace LIN
{
    public enum Type
    {
        Byte,       // 8-bit unsigned
        UInt16LE,   // 16-bit unsigned, little-endian (LSB first, MSB second)
        UInt16BE    // 16-bit unsigned, big-endian (MSB first, LSB second)
    }

    public class Parameter
    {

        public static int CountBytes(Type parameterType)
        {
            return parameterType switch
            {
                Type.Byte => 1,
                Type.UInt16LE => 2,
                Type.UInt16BE => 2,
                _ => throw new NotImplementedException(),
            };
        }

        // Format a parameter value from bytes to string for decompilation
        public static string FormatValue(Type parameterType, byte[] bytes, ref int byteIndex)
        {
            return parameterType switch
            {
                Type.Byte => bytes[byteIndex++].ToString(),
                Type.UInt16LE => FormatUInt16LE(bytes, ref byteIndex),
                Type.UInt16BE => FormatUInt16BE(bytes, ref byteIndex),
                _ => throw new NotImplementedException(),
            };
        }

        // Parse a parameter value from string to bytes for compilation
        public static void ParseValue(Type parameterType, string stringValue, List<byte> outputBytes, int lineNum)
        {
            try
            {
                switch (parameterType)
                {
                    case Type.Byte:
                        outputBytes.Add(byte.Parse(stringValue));
                        break;
                    case Type.UInt16LE:
                        ParseUInt16LE(stringValue, outputBytes);
                        break;
                    case Type.UInt16BE:
                        ParseUInt16BE(stringValue, outputBytes);
                        break;
                    default:
                        throw new NotImplementedException();
                }
            }
            catch (FormatException)
            {
                throw new Exception($"[read] error: invalid argument '{stringValue}' at line {lineNum + 1}");
            }
        }

        private static string FormatUInt16LE(byte[] bytes, ref int byteIndex)
        {
            // Little-endian: LSB first, MSB second
            ushort value = (ushort)(bytes[byteIndex] | (bytes[byteIndex + 1] << 8));
            byteIndex += 2;
            return value.ToString();
        }

        private static string FormatUInt16BE(byte[] bytes, ref int byteIndex)
        {
            // Big-endian: MSB first, LSB second
            ushort value = (ushort)((bytes[byteIndex] << 8) | bytes[byteIndex + 1]);
            byteIndex += 2;
            return value.ToString();
        }

        private static void ParseUInt16LE(string stringValue, List<byte> outputBytes)
        {
            ushort value = ushort.Parse(stringValue);
            // Little-endian: LSB first, MSB second
            outputBytes.Add((byte)(value & 0xFF));
            outputBytes.Add((byte)(value >> 8));
        }

        private static void ParseUInt16BE(string stringValue, List<byte> outputBytes)
        {
            ushort value = ushort.Parse(stringValue);
            // Big-endian: MSB first, LSB second
            outputBytes.Add((byte)(value >> 8));
            outputBytes.Add((byte)(value & 0xFF));
        }
    }
}

File: projects/lin-compiler/src/parameter.ts
/** Parameter encodings used by opcode arguments. */
export const ParamType = {
  /** 8-bit unsigned */
  Byte: "Byte",
  /** 16-bit unsigned, little-endian (LSB first, MSB second) */
  UInt16LE: "UInt16LE",
  /** 16-bit unsigned, big-endian (MSB first, LSB second) */
  UInt16BE: "UInt16BE",
} as const;
export type ParamType = (typeof ParamType)[keyof typeof ParamType];

/** Cursor into a byte array, so formatters can advance a shared read position. */
export interface ByteCursor {
  index: number;
}

export function countBytes(parameterType: ParamType): number {
  switch (parameterType) {
    case ParamType.Byte:
      return 1;
    case ParamType.UInt16LE:
    case ParamType.UInt16BE:
      return 2;
    default:
      throw new Error(`not implemented: ${parameterType}`);
  }
}

/** Format a parameter value from bytes to string for decompilation. */
export function formatValue(parameterType: ParamType, bytes: Uint8Array, cursor: ByteCursor): string {
  switch (parameterType) {
    case ParamType.Byte:
      return String(bytes[cursor.index++]);
    case ParamType.UInt16LE: {
      // Little-endian: LSB first, MSB second
      const value = bytes[cursor.index] | (bytes[cursor.index + 1] << 8);
      cursor.index += 2;
      return String(value);
    }
    case ParamType.UInt16BE: {
      // Big-endian: MSB first, LSB second
      const value = (bytes[cursor.index] << 8) | bytes[cursor.index + 1];
      cursor.index += 2;
      return String(value);
    }
    default:
      throw new Error(`not implemented: ${parameterType}`);
  }
}

/** Parse an unsigned integer, rejecting anything the C# `Parse` overloads would reject. */
export function parseUnsigned(stringValue: string, max: number): number {
  const trimmed = stringValue.trim();
  if (!/^\+?\d+$/.test(trimmed)) {
    throw new FormatError(stringValue);
  }
  const value = Number(trimmed);
  if (value > max) {
    throw new FormatError(stringValue);
  }
  return value;
}

/** Thrown for malformed numeric arguments; callers turn it into a line-annotated error. */
export class FormatError extends Error {
  readonly value: string;

  constructor(value: string) {
    super(`invalid numeric value '${value}'`);
    this.value = value;
    this.name = "FormatError";
  }
}

/** Parse a parameter value from string to bytes for compilation. */
export function parseValue(
  parameterType: ParamType,
  stringValue: string,
  outputBytes: number[],
  lineNum: number,
): void {
  try {
    switch (parameterType) {
      case ParamType.Byte:
        outputBytes.push(parseUnsigned(stringValue, 0xff));
        break;
      case ParamType.UInt16LE: {
        const value = parseUnsigned(stringValue, 0xffff);
        // Little-endian: LSB first, MSB second
        outputBytes.push(value & 0xff, value >> 8);
        break;
      }
      case ParamType.UInt16BE: {
        const value = parseUnsigned(stringValue, 0xffff);
        // Big-endian: MSB first, LSB second
        outputBytes.push(value >> 8, value & 0xff);
        break;
      }
      default:
        throw new Error(`not implemented: ${parameterType}`);
    }
  } catch (error) {
    if (error instanceof FormatError) {
      throw new Error(`[read] error: invalid argument '${stringValue}' at line ${lineNum + 1}`);
    }
    throw error;
  }
}
